vdr/ambassador.go

```diff
@@ -232,13 +232,29 @@ func (n *ambassador) handleUpdateDIDDocument(transaction dag.Transaction, propos
 		WithField(core.LogFieldTransactionRef, transaction.Ref()).
 		WithField(core.LogFieldDID, proposedDIDDocument.ID).
 		Debug("Handling DID document update")
-	// Resolve latest version of DID Document
-	currentDIDDocument, _, err := n.didStore.Resolve(proposedDIDDocument.ID, &types.ResolveMetadata{AllowDeactivated: true})
-	if err != nil {
-		return fmt.Errorf("unable to update DID document: %w", err)
+
+	// Resolve version of DID Document referred to by transaction
+	var currentDIDDocument *did.Document
+	var err error
+	for _, ref := range transaction.Previous() {
+		currentDIDDocument, _, err = n.didStore.Resolve(proposedDIDDocument.ID, &types.ResolveMetadata{AllowDeactivated: true, SourceTransaction: &ref})
+		if err != nil && !errors.Is(err, types.ErrNotFound) {
```

Comment:
Due to out of order event/notification processing the err could be a types.ErrNotFound even if the doc exists. So to answer the original question: yes, this could still be a recoverable error for the did update?

Reply:
There's still a fallback to find the latest. If the controller has been updated in the race condition this could go wrong, but there's no way to prevent that or fix it automatically (apart from the lock on API lvl which is already there).

Reply:
Exactly, that is what makes the error recoverable right? Since the order of processing cannot be guaranteed, an update that is invalid right now could be valid in the future.

Reply:
It's for controller finding. If the update succeeds with a controller from the previous/current state that's fine as well. If it fails, it might have been a race condition and then the retry will catch it.

Reply:
We are talking about the same thing here. Only recoverable errors are retried.